### PR TITLE
Refactoring langversion options in FSC

### LINKED ISSUE #13749: Refactor language version behavior in FSC
@KevinRansom Yep, thanks for the point, I think I will address that separately in a followup. That thing is just ridiculous - the function is called "setLanguageVersion" but actually it sometimes sets the version and sometimes gets the versions. That's like against all the coding principles, so I'll better do a small refactoring and testing around that.

_Originally posted by @psfinaki in https://github.com/dotnet/fsharp/issues/13702#issuecomment-1218054736_

### PR DESCRIPTION
**Pull request by psfinaki**
closes #13749

Splitting the all-in-one flag into two clear options, similar to how we have it with some other things like `target` or `debug`.

### CODE REVIEW

File: src/Compiler/Driver/CompilerOptions.fs [hunk 1]
```diff
@@ -1100,25 +1100,22 @@ let mlCompatibilityFlag (tcConfigB: TcConfigBuilder) =
...
+        sb.AppendLine v |> ignore
 
-        for v in languageVersion.ValidOptions do
-            printfn "%s" v
+    for v in LanguageVersion.Default.ValidVersions do
+        sb.AppendLine v |> ignore
 
-        for v in languageVersion.ValidVersions do
-            printfn "%s" v
+    sb.ToString()
```
Comment (0101) [inline src/Compiler/Driver/CompilerOptions.fs:1113]: How about
```fsharp
seq { 
    FSComp.SR.optsSupportedLangVersions ()
    yield! LanguageVersion.Default.ValidOptions
    yield! LanguageVersion.Default.ValidVersions
} |> String.concat "\n"

**Review by 0101 — APPROVED** [review]

**Review by KevinRansom — CHANGES_REQUESTED** [review]
We should talk

File: src/Compiler/Driver/CompilerOptions.fs (hunk 1 above)
Comment (psfinaki) [reply src/Compiler/Driver/CompilerOptions.fs]: Yep, good idea, changed

File: src/Compiler/Driver/CompilerOptions.fs [hunk 2]
```diff
@@ -1100,25 +1100,22 @@ let mlCompatibilityFlag (tcConfigB: TcConfigBuilder) =
...
+    sb.AppendLine(FSComp.SR.optsSupportedLangVersions ()) |> ignore
 
-    let dumpAllowedValues () =
-        printfn "%s" (FSComp.SR.optsSupportedLangVersions ())
+    for v in LanguageVersion.Default.ValidOptions do
+        sb.AppendLine v |> ignore
 
-        for v in languageVersion.ValidOptions do
-            printfn "%s" v
+    for v in LanguageVersion.Default.ValidVersions do
```
[the remark this answers is not in the document]
Comment (psfinaki) [reply src/Compiler/Driver/CompilerOptions.fs]: Rebased on that

File: src/Compiler/Driver/CompilerOptions.fs [hunk 3]
```diff
@@ -1127,18 +1124,28 @@ let setLanguageVersion specifiedVersion =
...
         //                               'default' (latest major version), or
         //                               'latest' (latest version, including minor versions),
         //                               'preview' (features for preview)
         //                               or specific versions like '4.7'
+        CompilerOption(
+            "langversion",
+            tagNone,
+            OptionConsoleOnly(fun _ ->
+                Console.Write(GetLanguageVersions())
+                QuitProcessExiter.Exit 0),
```
[the remark this answers is not in the document]
Comment (psfinaki) [reply src/Compiler/Driver/CompilerOptions.fs]: Returned exit 0 for now

File: src/Compiler/Driver/CompilerOptions.fsi [hunk 4]
```diff
@@ -58,6 +58,8 @@ val GetHelpFsc: tcConfigB: TcConfigBuilder -> blocks: CompilerOptionBlock list -
 
 val GetVersion: tcConfigB: TcConfigBuilder -> string
 
+val GetLanguageVersions: unit -> string
```
[the remark this answers is not in the document]
Comment (psfinaki) [reply src/Compiler/Driver/CompilerOptions.fsi:61]: For testing

**Pull request closed without merging**